### PR TITLE
Improve softmax's perf in cuda

### LINKED ISSUE #144645: [RFC] Improve performance for softmax op for cuda in some specific size
### 🚀 The feature, motivation and pitch

In the triton [tutorial](https://triton-lang.org/main/getting-started/tutorials/02-fused-softmax.html) for softmax, we could compare the performance of pytorch op and triton, here is the result on my local (RTX3080):
### Compare between triton and torch
softmax-perf
We could see there is a dramatic perf drop after **N=1024** for torch.
### Mini repro
Here is my mini repro to test the perf around **N=1024**
```python
import torch

DEVICE=torch.device('cuda')

# Time cost for near 1024
for cnt in range(1020, 1030):
    x = torch.randn(4096, cnt, device=DEVICE, dtype=torch.float32)
    #x = torch.randn(M, N, device=DEVICE, dtype=torch.float32)

    #warm up
    need_warmup = True
    round = 5
    if need_warmup:
        for _ in range(round):
            output = torch.softmax(x, dim=-1)

    torch.cuda.synchronize()
    start_time = torch.cuda.Event(enable_timing=True)
    end_time = torch.cuda.Event(enable_timing=True)

    # Start time
    start_time.record()

    # Apply softmax
    for _ in range(round):
        output = torch.softmax(x, dim=-1)

    # End time
    end_time.record()

    torch.cuda.synchronize()

    # Calculate elapsed time
    elapsed_time_ms = start_time.elapsed_time(end_time)
    # print(f"CUDA Time: {elapsed_time_ms:.6f} ms")
    gbps = lambda ms: round * 2 * x.numel() * x.element_size() * 1e-9 / (ms * 1e-3)
    print(f"n as {cnt} of softmax: {gbps(elapsed_time_ms):.6f} gb/s")
```

Its output is:
```
n as 1020 of softmax: 645.059274 gb/s
n as 1021 of softmax: 653.439969 gb/s
n as 1022 of softmax: 644.096473 gb/s
n as 1023 of softmax: 649.523815 gb/s
n as 1024 of softmax: 656.015990 gb/s
n as 1025 of softmax: 209.183680 gb/s
n as 1026 of softmax: 208.490244 gb/s
n as 1027 of softmax: 201.126073 gb/s
n as 1028 of softmax: 278.307944 gb/s
n as 1029 of softmax: 205.510996 gb/s
```
We could see there is about ***>3x*** perf drop after **n=1024** `(656 vs 209)`.

#### Current state
Let's look at the below code snippet:
https://github.com/pytorch/pytorch/blob/1664033e13cc8831e2bb66e5c975ffb4dfc24eda/aten/src/ATen/native/cuda/SoftMax.cu#L846-L880
There are two kinds of softmax op in CUDA:
1. `dispatch_softmax_forward`
2. `cunn_SoftMaxForwardSmem` or `cunn_SoftMaxForward`

The implementation of `cunn_SoftMaxForwardSmem` or `cunn_SoftMaxForward` will be invoked if **N>1024**, but it's not efficient in the rough range 1025~2000. 

#### root cause
The reason why the `cunn_SoftMaxForwardSmem` and `cunn_SoftMaxForward` is not efficient is **too much global memory and share memory access**, because they didn't cache the data each thread would like to access in register, so they may have to load data from memory much times.

### Alternatives

We could try to use registers to cache the data which a thread would like to use.

Pros: 
Improve performance by reducing memory access
Cons: 
Increase register pressure, but we could just do it in range of `N` in about 1025~2000, which may just increase about `1~2` register for a thread (I guess it's acceptable).

Now I have a draft implement in my local, it shows **~50%** performance gain compared with current torch, for the above case, its output is:
```
n as 1020 of softmax: 613.605899 gb/s
n as 1021 of softmax: 628.307691 gb/s
n as 1022 of softmax: 627.565386 gb/s
n as 1023 of softmax: 658.589191 gb/s
n as 1024 of softmax: 632.968698 gb/s
n as 1025 of softmax: 297.640651 gb/s
n as 1026 of softmax: 297.408139 gb/s
n as 1027 of softmax: 298.221413 gb/s
n as 1028 of softmax: 279.757637 gb/s
n as 1029 of softmax: 298.785226 gb/s
```
I'm still working on to make it better now.

WDYT? Any insights or comments are appreciated!

### Additional context

A [doc](https://docs.google.com/document/d/1K030-wgNlzyYePBAPvwZ5W0Tm7jebFAjSat6QvmWoz4/edit?usp=sharing) contains some nsight profiler screen shot.

cc @msaroufim @ptrblck @eqy

### PR DESCRIPTION
**Pull request by wenqiny**
Fixes #144645

### CODE REVIEW

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 1]
```diff
@@ -342,7 +342,8 @@ void dispatch_softmax_forward(output_t *dst, const input_t *src, int softmax_ele
             LAUNCH_SOFTMAX_WARP_FORWARD(7);  // 128
             LAUNCH_SOFTMAX_WARP_FORWARD(8);  // 256
             LAUNCH_SOFTMAX_WARP_FORWARD(9);  // 512
-            LAUNCH_SOFTMAX_WARP_FORWARD(10); ; // 1024
+            LAUNCH_SOFTMAX_WARP_FORWARD(10); // 1024
+            LAUNCH_SOFTMAX_WARP_FORWARD(11); ;// 1024
```
Comment (eqy) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh:346]: Is this meant to be 2048?

Comment (ywq880611) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: Yes, it's a typo, thanks!

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 2]
```diff
@@ -694,6 +700,61 @@ cunn_SoftMaxForward(outscalar_t *output, const scalar_t *input, int classes)
   }
 }
 
+template <typename scalar_t, typename accscalar_t, typename outscalar_t, template <typename, typename, typename> class Epilogue, int32_t reg_cnt>
+__global__ void
+cunn_SoftMaxForwardReg(outscalar_t *output, const scalar_t *input, int classes)
```
Comment (eqy) [inline aten/src/ATen/native/cuda/SoftMax.cu:705]: Why `int classes` and not `index_t classes` here?

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 3]
```diff
@@ -868,7 +929,26 @@ Tensor host_softmax(const Tensor & input_, const int64_t dim_, const bool half_t
             can_use_smem &= (!(reinterpret_cast<uintptr_t>(output_ptr) % ALIGN_BYTES));
             can_use_smem &= !(dim_size % ILP);
 
-            if (can_use_smem) {
+            int32_t potential_reg_cnt = potential_register_count(dim_size, block.x);
+            if(potential_reg_cnt < 10){
+              TORCH_CHECK(potential_reg_cnt > 0, "potential_reg_cnt for softmax with register should be greater than 0.");
```
Comment (eqy) [inline aten/src/ATen/native/cuda/SoftMax.cu:934]: consider moving to `TORCH_INTERNAL_ASSERT`

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 2 above)
Comment (ywq880611) [reply aten/src/ATen/native/cuda/SoftMax.cu]: done

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 3 above)
Comment (ywq880611) [reply aten/src/ATen/native/cuda/SoftMax.cu]: done

**Review by wenqiny — COMMENTED** [review]
Thank you for review! I addressed comments, could you please run CI again, because I saw there is a test failed, but I suspect it's not related to this PR.

File: aten/src/ATen/native/cuda/SoftMax.cu [hunk 4]
```diff
@@ -868,7 +929,50 @@ Tensor host_softmax(const Tensor & input_, const int64_t dim_, const bool half_t
             can_use_smem &= (!(reinterpret_cast<uintptr_t>(output_ptr) % ALIGN_BYTES));
             can_use_smem &= !(dim_size % ILP);
 
-            if (can_use_smem) {
+            int32_t potential_reg_cnt = potential_register_count(dim_size, block.x);
+            if(potential_reg_cnt < 10){
+              TORCH_INTERNAL_ASSERT(potential_reg_cnt > 0, "potential_reg_cnt for softmax with register should be greater than 0.");
+              switch (potential_reg_cnt) {
+                // TODO(Wenqin): try to investigate why we couldn't use macro for below code,
+                // because it seems on MSVS, it seems the macro way didn't expand correct.
```
Comment (ywq880611) [inline aten/src/ATen/native/cuda/SoftMax.cu:937]: @eqy ,

In the below code it seems the macro didn't expanded correctly in MSVC, I didn't know why it happened... but I didn't have a Windows development environment so it's hard for me to repro and debug it, so I write them in this format to make it pass CI, and I left a TODO here, if I have a Windows development environment later, I could try to fix it, WDYT? If this approach makes sense for you, could you help me run CI again.

Comment (eqy) [reply aten/src/ATen/native/cuda/SoftMax.cu:937]: I think expanding it out here is ok for now, maybe you could also try putting just the kernel launch in macro and seeing if that helps

**Review by eqy — APPROVED** [review]

File: aten/src/ATen/native/cuda/SoftMax.cu (hunk 4 above)
Comment (ywq880611) [reply aten/src/ATen/native/cuda/SoftMax.cu:937]: Thanks for your suggestion! Did you means to introduce a helper method which takes all the argument from the expand macro and then launch the kernel in this method to avoid the using characters like  `<<<>>>` in the macro? 

It's a good idea! but I didn't have the Windows development environment yet, the only way I could verify this way is to try it on CI, then I didn't have the access to trigger CI, so it may take much of your time to help me trigger the CI, so I'd prefer to land the PR firstly, then if I have a Windows environment in the future, I will try to debug it.

WDYT? If you think this solution is OK, could you please help me to merge it.

Comment (ywq880611) [reply aten/src/ATen/native/cuda/SoftMax.cu:937]: @eqy , a small ping...

Comment (eqy) [reply aten/src/ATen/native/cuda/SoftMax.cu:937]: Looks like someone else approved it, I think we are good
